lib/evaluate/shape.h

```diff
@@ -86,56 +86,57 @@ MaybeExtentExpr GetSize(Shape &&);
 // Utility predicate: does an expression reference any implied DO index?
 bool ContainsAnyImpliedDoIndex(const ExtentExpr &);
 
-// Compilation-time shape conformance checking, when corresponding extents
-// are known.
-bool CheckConformance(parser::ContextualMessages &, const Shape &,
-    const Shape &, const char * = "left operand",
-    const char * = "right operand");
-
-class GetShapeVisitor : public virtual VisitorBase<std::optional<Shape>> {
+// GetShape()
+class GetShapeHelper
+  : public AnyTraverse<GetShapeHelper, std::optional<Shape>> {
 public:
   using Result = std::optional<Shape>;
-  explicit GetShapeVisitor(FoldingContext &c) : context_{c} {}
+  using Base = AnyTraverse<GetShapeHelper, Result>;
+  using Base::operator();
+  GetShapeHelper(FoldingContext &c) : Base{*this}, context_{c} {}
+
+  Result operator()(const ImpliedDoIndex &) const { return Scalar(); }
+  Result operator()(const DescriptorInquiry &) const { return Scalar(); }
+  template<int KIND> Result operator()(const TypeParamInquiry<KIND> &) const {
+    return Scalar();
+  }
+  Result operator()(const BOZLiteralConstant &) const { return Scalar(); }
+  Result operator()(const StaticDataObject::Pointer &) const {
+    return Scalar();
+  }
+  Result operator()(const StructureConstructor &) const { return Scalar(); }
 
-  template<typename T> void Handle(const Constant<T> &c) {
-    Return(AsShape(c.SHAPE()));
+  template<typename T> Result operator()(const Constant<T> &c) const {
+    return AsShape(c.SHAPE());
   }
-  void Handle(const Symbol &);
-  void Handle(const Component &);
-  void Handle(const NamedEntity &);
-  void Handle(const StaticDataObject::Pointer &) { Scalar(); }
-  void Handle(const ArrayRef &);
-  void Handle(const CoarrayRef &);
-  void Handle(const Substring &);
-  void Handle(const ProcedureRef &);
-  void Handle(const StructureConstructor &) { Scalar(); }
-  template<typename T> void Handle(const ArrayConstructor<T> &aconst) {
-    Return(Shape{GetArrayConstructorExtent(aconst)});
+
+  Result operator()(const Symbol &) const;
+  Result operator()(const Component &) const;
+  Result operator()(const NamedEntity &) const;
+  Result operator()(const ArrayRef &) const;
+  Result operator()(const CoarrayRef &) const;
+  Result operator()(const Substring &) const;
+  Result operator()(const ProcedureRef &) const;
+
+  template<typename T>
+  Result operator()(const ArrayConstructor<T> &aconst) const {
+    return Shape{GetArrayConstructorExtent(aconst)};
   }
-  void Handle(const ImpliedDoIndex &) { Scalar(); }
-  void Handle(const DescriptorInquiry &) { Scalar(); }
-  template<int KIND> void Handle(const TypeParamInquiry<KIND> &) { Scalar(); }
-  void Handle(const BOZLiteralConstant &) { Scalar(); }
-  void Handle(const NullPointer &) { Return(); }
   template<typename D, typename R, typename LO, typename RO>
-  void Handle(const Operation<D, R, LO, RO> &operation) {
+  Result operator()(const Operation<D, R, LO, RO> &operation) const {
```

Comment:
As mentioned more in detail in another comment, I do not think this overload can be selected.
See:
```
  integer :: x, y(10)
  integer, parameter :: p1(*) = shape(abs(x) + y), p2(*) = shape(y + abs(x))
end
```

Reply:
This example works now.

Reply:
This case works now.